File: NodeGarden/NodeGardenDirect3D/NodeGardenDirect3DComp/LineConnection.cpp
#include "pch.h"
#include "LineConnection.h"

const float LineConnection::StrokeWeightMin = 2.0f; 
const float LineConnection::StrokeWeightMax = 7.0f;

LineConnection::LineConnection()
{
    m_zDepth = 1.0f;
}

void LineConnection::FormConnection(XMVECTOR node1Pos, XMVECTOR node2Pos, float distance)
{
    m_visible = true;
    // draw a line between 2 nodes. The thickness/alpha varies depending on distance
    m_strokeThickness = Node::Map(distance, 0, MinDist, StrokeWeightMax, StrokeWeightMin);
    
    XMVECTORF32 color = {1, 1, 1, Node::Map(distance, 0, MinDist, 1.0f, 0)};
    m_color = color;

    XMStoreFloat2(&m_start, node1Pos);
    XMStoreFloat2(&m_end, node2Pos);

    m_destRect.left =   (long)m_start.x;
    m_destRect.top =    (long)m_start.y;
    m_destRect.right  = (long)(m_start.x + m_strokeThickness);  
    m_destRect.bottom = (long)(m_start.y + Distance(XMLoadFloat2(&m_start), XMLoadFloat2(&m_end)));

    m_rotation = PIOVER2 - (float)atan2(m_end.y - m_start.y, m_start.x - m_end.x);
}

void LineConnection::BreakConnection()
{
    m_visible = false;
}

float LineConnection::Distance(const XMVECTOR& vector1,const XMVECTOR& vector2)
{
    XMVECTOR vectorSub = XMVectorSubtract(vector1,vector2);
    XMVECTOR length = XMVector3Length(vectorSub);

    float distance = 0.0f;
    XMStoreFloat(&distance,length);
    return distance;
}

File: NodeGarden/NodeGardenXNA/NodeGardenXNA/LineConnection.cs
//-----------------------------------------------------------------------
// <copyright file="LineConnection.cs" company="Studio Arcade Ltd">
// Copyright © Studio Arcade Ltd 2012.
// All rights reserved.
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY 
// KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
//
// This code is made available under the Ms-PL or GPL as appropriate.
// Please see LICENSE.txt for more details
// </copyright>
// <Author>Matt Lacey</Author>
// <Author>Laurie Brown</Author>
//-----------------------------------------------------------------------

namespace NodeGardenXNA
{
    using System;

    using Microsoft.Xna.Framework;
    using Microsoft.Xna.Framework.Graphics;

    /// <summary>
    /// A line connecting nodes
    /// </summary>
    public class LineConnection : Connection
    {
        /// <summary>
        /// Minimum stroke weight for a connecting line
        /// </summary>
        protected const float StrokeWeightMin = 1.0f;

        /// <summary>
        /// Maximum stroke weight for a connecting line
        /// </summary>
        protected const float StrokeWeightMax = 7.0f;

        /// <summary>
        /// The texture to draw
        /// </summary>
        private readonly Texture2D pixelTexture;

        /// <summary>
        /// Actual stroke thickness of connecting line
        /// </summary>
        private int strokeThickness;

        /// <summary>
        /// Starting position
        /// </summary>
        private Vector2 start;

        /// <summary>
        /// Ending position
        /// </summary>
        private Vector2 end;

        /// <summary>
        /// Color of the line
        /// </summary>
        private Color color;

        /// <summary>
        /// Is this line visible
        /// </summary>
        private bool isVisible;

        /// <summary>
        /// Initializes a new instance of the <see cref="LineConnection"/> class.
        /// </summary>
        /// <param name="pixelTexture">The pixel texture.</param>
        public LineConnection(Texture2D pixelTexture)
        {
            this.pixelTexture = pixelTexture;
        }

        /// <summary>
        /// Forms the connection.
        /// </summary>
        /// <param name="node1">The node1.</param>
        /// <param name="node2">The node2.</param>
        /// <param name="distance">The distance.</param>
        public override void FormConnection(VisualNode node1, VisualNode node2, float distance)
        {
            this.isVisible = true;

            // draw a line between 2 nodes. The thickness/alpha varies depending on distance
            this.strokeThickness = (int)Global.Map(distance, 0, Global.MinDist, StrokeWeightMax, StrokeWeightMin);
            this.color = Color.White * Global.Map(distance, 0, Global.MinDist, 1.0f, 0);
            this.start = node1.CurrentPosition;
            this.end = node2.CurrentPosition;
        }

        /// <summary>
        /// Breaks the connection.
        /// </summary>
        /// <param name="node1">The node1.</param>
        /// <param name="node2">The node2.</param>
        public override void BreakConnection(VisualNode node1, VisualNode node2)
        {
            this.isVisible = false;
        }

        /// <summary>
        /// Draws the specified sprite batch.
        /// </summary>
        /// <param name="sb">The sprite batch.</param>
        public override void Draw(SpriteBatch sb)
        {
            if (this.isVisible)
            {
                sb.Draw(
                    this.pixelTexture,
                    new Rectangle(
                        (int)this.start.X,
                        (int)this.start.Y,
                        this.strokeThickness,
                        (int)Vector2.Distance(this.start, this.end)),
                    null,
                    this.color,
                    MathHelper.PiOver2 - (float)Math.Atan2(this.end.Y - this.start.Y, this.start.X - this.end.X),
                    Vector2.Zero,
                    SpriteEffects.None,
                    0.5f);
            }
        }
    }
}